src/functions/casing.ts

```diff
@@ -1,40 +1,95 @@
-  flat: /^[a-z][a-z0-9]*$/,
-  camel: /^[a-z][a-z0-9]*(?:[A-Z0-9][a-z0-9]+)*$/,
-  pascal: /^[A-Z][a-z0-9]*(?:[A-Z0-9][a-z0-9]+)*$/,
-  kebab: /^[a-z][a-z0-9]*(?:-[a-z0-9]+)*$/,
-  cobol: /^[A-Z][A-Z0-9]*(?:-[A-Z0-9]+)*$/,
-  snake: /^[a-z][a-z0-9]*(?:_[a-z0-9]+)*$/,
-  macro: /^[A-Z][A-Z0-9]*(?:_[A-Z0-9]+)*$/,
+export type CasingRule = IRule<RuleFunction.CASING, ICasingOptions>;
+
+const CASES: Dictionary<string, CasingType> = {
+  [CasingType.flat]: '[a-z][a-z{__DIGITS__}]*',
+  [CasingType.camel]: '[a-z][a-z{__DIGITS__}]*(?:[A-Z{__DIGITS__}][a-z{__DIGITS__}]+)*',
+  [CasingType.pascal]: '[A-Z][a-z{__DIGITS__}]*(?:[A-Z{__DIGITS__}][a-z{__DIGITS__}]+)*',
+  [CasingType.kebab]: '[a-z][a-z{__DIGITS__}]*(?:-[a-z{__DIGITS__}]+)*',
+  [CasingType.cobol]: '[A-Z][A-Z{__DIGITS__}]*(?:-[A-Z{__DIGITS__}]+)*',
+  [CasingType.snake]: '[a-z][a-z{__DIGITS__}]*(?:_[a-z{__DIGITS__}]+)*',
+  [CasingType.macro]: '[A-Z][A-Z{__DIGITS__}]*(?:_[A-Z{__DIGITS__}]+)*',
 };
 
 export const casing: IFunction<ICasingOptions> = (targetVal, opts) => {
   if (typeof targetVal !== 'string' || targetVal.length === 0) {
     return;
   }
 
-  const set = opts.disallowDigits ? CASES : CASES_WITH_DIGITS;
+  if (!validate(opts)) {
+    return;
+  }
+
+  const casingValidator = buildFrom(CASES[opts.type], opts);
+
+  if (casingValidator.test(targetVal)) {
+    return;
+  }
+
+  return [
+    {
+      message: `must be ${opts.type} case`,
+    },
+  ];
+};
+
+const validate = (opts: ICasingOptions): boolean => {
+  if (!(opts.type in CASES)) {
+    // TODO: throw a relevant error message
+    return false;
+  }
+
+  if (opts.allowLeadingSeparator !== undefined && opts.separator === undefined) {
+    throw new Error("'allowLeadingSeparator' can only be valued when 'separator' is defined.");
+  }
 
-  if (opts.type in set && !set[opts.type].test(targetVal)) {
-    return [
-      {
-        message: `must be ${opts.type} case`,
-      },
-    ];
+  if (opts.separator !== undefined && opts.separator.length !== 1) {
+    throw new Error("When valued, 'separator' only accepts a single char.");
   }
 
-  return;
+  return true;
 };
+
+const DIGITS_PATTERN = '0-9';
+
+const buildFrom = (basePattern: string, overrides: ICasingOptionsOverrides): RegExp => {
+  const injectDigits = overrides.disallowDigits === undefined || overrides.disallowDigits === false;
+
+  const pattern = basePattern.replace(/\{__DIGITS__\}/g, injectDigits ? DIGITS_PATTERN : '');
+
+  if (overrides.separator === undefined) {
+    return new RegExp(`^${pattern}$`);
+  }
+
+  const separatorPattern = `[${escapeRegExp(overrides.separator)}]`;
+  const leadingSeparatorPattern = overrides.allowLeadingSeparator === true ? `${separatorPattern}?` : '';
+
+  return new RegExp(`^${leadingSeparatorPattern}${pattern}(?:${separatorPattern}${pattern})*$`);
+};
+
+// From https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions
+function escapeRegExp(string: string) {
```

Comment:
You can use function exported by lodash.

Reply:
Fixed